src/Raven.Server/Documents/Indexes/Static/DynamicArray.cs

```diff
@@ -105,6 +105,11 @@ public override bool TryConvert(ConvertBinder binder, out object result)
             return base.TryConvert(binder, out result);
         }
 
+        public dynamic this[int i]
+        {
+            get { return ElementAt(i); }
```

Comment:
Something like this (haven't tested).

```suggestion
            get { 
                    if(_inner is not ICollection c) {
                         _inner = c = _inner .ToList();
                    }
                    return c[i];                    
            }
```

Reply:
Looks like ElementAt is doing same thing: https://github.com/dotnet/runtime/blob/main/src%2Flibraries%2FSystem.Linq%2Fsrc%2FSystem%2FLinq%2FElementAt.cs#L12 ?

Reply:
Yes, but our _enumerable_ is often _not_ an `ICollection`, it is a `items.Select(x =>)`.
And then, when you call `[2]` on that, you materialize all pervious values.
In a `for loop`, you'll do that `N^2` times.

The code above would take the usage of the indexer as a hint, materialize the values once and then access them by index.

Reply:
For that matter, we want to do that for `Count` as well

Reply:
There are other places in the very same file that use ElementAt. At this point I do not consider this a problem. If needed we can optimize the whole class in future.

Reply:
https://issues.hibernatingrhinos.com/issue/RavenDB-18158